include/fw_metadata.h

```diff
@@ -0,0 +1,241 @@
+ * firmware. It will also likely make it easier to add all the info at compile
+ * time.
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <toolchain.h>
+#include <assert.h>
+#include <string.h>
+#include <debug.h>
+
+#define MAGIC_LEN_WORDS (CONFIG_FW_MAGIC_LEN / sizeof(u32_t))
+
+struct fw_abi_info;
+
+/**@brief Function that returns an ABI.
+ *
+ * @param[in]    id      Which ABI to get.
+ * @param[in]    index   If there are multiple ABIs available with the same ID,
+ *                       retrieve the different ones with this.
+ * @param[out]   abi     Pointer to the abi with the given id and index.
+ *
+ * @retval 0        Success.
+ * @retval -ENOENT  id not found.
+ * @retval -EBADF   index too large.
+ * @retval -EFAULT  abi was NULL.
+ */
+typedef int (*fw_abi_getter)(u32_t id, u32_t index,
+				const struct fw_abi_info **abi);
+
+struct __packed fw_firmware_info {
+	/* Magic value to verify that the struct has the correct type. */
+	u32_t magic[MAGIC_LEN_WORDS];
+
+	/* Size without validation_info pointer and padding. */
+	u32_t firmware_size;
+
+	/* Monotonically increasing version counter.*/
+	u32_t firmware_version;
+
+	/* The address of the start (vector table) of the firmware. */
+	u32_t firmware_address;
+
+	/* Where to place the getter for the ABI provided to this firmware. */
+	fw_abi_getter *abi_in;
+
+	/* This firmware's ABI getter. */
+	const fw_abi_getter abi_out;
+};
+
+
+#define OFFSET_CHECK(type, member, value) \
+		BUILD_ASSERT_MSG(offsetof(type, member) == value, \
+				#member " has wrong offset")
+
+/* Static asserts to ensure compatibility */
+OFFSET_CHECK(struct fw_firmware_info, magic, 0);
+OFFSET_CHECK(struct fw_firmware_info, firmware_size, CONFIG_FW_MAGIC_LEN);
+OFFSET_CHECK(struct fw_firmware_info, firmware_version,
+	(CONFIG_FW_MAGIC_LEN + 4));
+OFFSET_CHECK(struct fw_firmware_info, firmware_address,
+	(CONFIG_FW_MAGIC_LEN + 8));
+
+/* For declaring this firmware's firmware info. */
+#define __fw_info Z_GENERIC_SECTION(.firmware_info) __attribute__((used)) const
+
+/* This struct is meant to serve as a header before a list of function pointers
+ * (or something else) that constitute the actual ABI. How to use the ABI, such
+ * as the signatures of all the functions in the list must be unambiguous for an
+ * ID/version combination.
+ */
+struct __packed fw_abi_info {
+	/* Magic value to verify that the struct has the correct type. */
+	u32_t magic[MAGIC_LEN_WORDS];
+
+	/* The id of the ABI. */
+	u32_t abi_id;
+
+	/* Flags specifying properties of the ABI. */
+	u32_t abi_flags;
+
+	/* The version of this ABI. */
+	u32_t abi_version;
+
+	/* The length of this header plus everything after this header. Must be
+	 * word-aligned.
+	 */
+	u32_t abi_len;
+};
+
+
+#define OFFSET_CHECK_EXT_ABI(type, member, value) \
+	BUILD_ASSERT_MSG(offsetof(type, header.member) == value, \
+		"ext_abi " #type " has wrong offset for header." #member)
+
+#define __ext_abi(type, name) \
```

Comment:
Don't care much here, but macros should be uppercase and not `__` prefixed since those symbols are, in theory, reserved for linker symbols. They have started to enforce this upstream because of MISRA-C compliance.

Reply:
It's named like this because it's meant to be used the same way as e.g. `__noinit`

Reply:
Ok, thanks.